homeassistant/components/imap/sensor.py

```diff
@@ -54,139 +46,60 @@ async def async_setup_platform(
-        if self._connection is None:
-            try:
-                self._connection = IMAP4_SSL(self._server, self._port)
-                await self._connection.wait_hello_from_server()
-                await self._connection.login(self._user, self._password)
-                await self._connection.select(self._folder)
-                self._does_push = self._connection.has_capability("IDLE")
-            except (AioImapException, asyncio.TimeoutError):
-                self._connection = None
-
-        return self._connection
-
-    async def idle_loop(self):
-        """Wait for data pushed from server."""
-        while True:
-            try:
-                if await self.connection():
-                    await self.refresh_email_count()
-                    self.async_write_ha_state()
-
-                    idle = await self._connection.idle_start()
-                    await self._connection.wait_server_push()
-                    self._connection.idle_done()
-                    async with async_timeout.timeout(10):
-                        await idle
-                else:
-                    self.async_write_ha_state()
-            except (AioImapException, asyncio.TimeoutError):
-                self.disconnected()
+    def native_value(self) -> int:
+        """Return the number of emails found."""
+        return self.coordinator.data
 
     async def async_update(self) -> None:
-        """Periodic polling of state."""
-        try:
-            if await self.connection():
-                await self.refresh_email_count()
-        except (AioImapException, asyncio.TimeoutError):
-            self.disconnected()
-
-    async def refresh_email_count(self):
-        """Check the number of found emails."""
-        if self._connection:
-            await self._connection.noop()
-            result, lines = await self._connection.search(
-                self._search, charset=self._charset
-            )
-
-            if result == "OK":
-                self._email_count = len(lines[0].split())
-            else:
-                _LOGGER.error(
-                    "Can't parse IMAP server response to search '%s':  %s / %s",
-                    self._search,
-                    result,
-                    lines[0],
-                )
-
-    def disconnected(self):
-        """Forget the connection after it was lost."""
-        _LOGGER.warning("Lost %s (will attempt to reconnect)", self._server)
-        self._connection = None
-
-    async def shutdown(self, *_):
-        """Close resources."""
-        if self._connection:
-            if self._connection.has_pending_idle():
-                self._connection.idle_done()
-            await self._connection.logout()
-        if self._idle_loop_task:
-            self._idle_loop_task.cancel()
+        """Check for idle state before updating."""
+        if not await self.coordinator.imap_client.stop_wait_server_push():
```

Comment:
Is the check here to not request double refreshes?

Reply:
Yup, if the IMAP server doesn't support idling then we need to request refresh manually.